<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment</b>
    
Hello, my name is Sveta and I am going to review this project. 

Before we start, I want to pay your attention to the color marking:
    
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Great solutions and ideas that can and should be used in the future are in green comments.   
</div>    
    
    
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>

Yellow color indicates what should be optimized. This is not necessary, but it will be great if you make changes to this project.
</div>      
    
    
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>

Issues that need to be corrected to get right results are indicated in red comments. Note that the project cannot be accepted until these issues are resolved.
</div>    

<hr>
    
**Please, use some color other than those listed to highlight answers to my comments.**
I would also ask you **not to change, move or delete my comments** so that it would be easier for me to navigate during the next review.
    
In addition, my comments are defined as headings. 
They can mess up the content, however, they are convenient, since you can immediately go to them. I will remove the headings from my comments in the next review. 
   
    
    
<hr>
    
<font color='dodgerblue'>**A few words about the project:**</font> you did a great job, everything is clear and neat. However, there are some issues that need to be fixed. I've also  left some comments with the recommendations for improving the project. If you have any questions, feel free to ask me.
    
    
I will wait for the project for a second review :)
    
    
📌 Here are some tips and tricks that may help you with the Markdown cells:    
<hr style="border-top: 3px solid purple; "></hr>

You can leave comments using this code inside a Markdown cell:
    
    
    <div class="alert alert-info">
    <h2> Student's comment</h2>

    Your text here. 
    </div>

    
    
    <font color='red'> This code is used to change text color. </font>     

<font color='red'> It will look like this. </font> 
    
If you don't want your comments to be headings, replace **h2** with **b** or just add `<a class="tocSkip">` after the phrase *Student's comment*.


You can find out how to **format text** in a Markdown cell or how to **add links** [here](https://sqlbak.com/blog/jupyter-notebook-markdown-cheatsheet) и [and here](https://medium.com/analytics-vidhya/the-ultimate-markdown-guide-for-jupyter-notebook-d5e5abf728fd).



</div>

<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 2</b>


Thank you for correcting your project! :) I've left new comments titled as **Reviewer's comment 2**. Please take a look, it should not take much time. 
    
   
</div>

<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 3</b>


My new comments are titled as **Reviewer's comment 3**. Your project has passed code review. Congratulations 😊
 
    
This is a good [article](https://www.kaggle.com/ramprakasism/pandas-75-exercises-with-solutions/notebook) with 75 pandas exercises and solutions. You can go along the exercises and compare your solution with the solution in the article.
    
    
    
Take care and good luck! 😊      
    
    
    
<hr>    
    
    
Best regards, 
    
    
S.N.    
</div>

## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    
There's an introduction, which is good. It is important to write an introductory part because it gives an idea about the content of the project. Don't forget about it in the next projects 😊
    
    
</div>

## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [1]:
# importing pandas
import pandas as pd

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    
It is good that you are using **pd** as a reference to pandas. It is convenient as every programmer or analyst has to look for the information on the web or in a documentation. Sometimes you may search for a ready solution and use of a common style is a good idea here as it simplifies understanding someone's code and even copy-paste process. 
     
</div>

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [2]:
# reading the files and storing them to df
df = pd.read_csv('/datasets/movies_and_shows.csv')


Print the first 10 table rows:

In [3]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
display(df.head(10))
#displaying table rows correction 

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment </h2>
    
It's much better to use display instead of print for dataframes. Take a look:     
</div>

In [4]:
# Reviewer's code

display(df.head())

name                Character   r0le        TITLE   Type  \
0   Robert De Niro            Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster            Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                      Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                    Betsy  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment </h2>
    
Display is a great thing. The point is, jupyter notebook is an interactive environment that already includes display. When we call a dataframe (see the code below), jupyter prints this dataframe like we do with the display method:</div>

In [5]:
# Reviewer's code

df.head()

name                Character   r0le        TITLE   Type  \
0   Robert De Niro            Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster            Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                      Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                    Betsy  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment </h2>
    
If you want to print more than one dataframe, only last command will be executed: 

</div>

In [6]:
# Reviewer's code

df.head(4)
df.head(3)
df.head(2)

name      Character   r0le        TITLE   Type  release Year  \
0  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver  MOVIE          1976   
1    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver  MOVIE          1976   

               genres  imdb sc0re  imdb v0tes  
0  ['drama', 'crime']         8.2    808582.0  
1  ['drama', 'crime']         8.2    808582.0

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment </h2>
    
And that's where we need **display**: </div>

In [7]:
# Reviewer's code

display(df.head(4))
display(df.head(3))

df.head(2)

name                Character   r0le        TITLE   Type  \
0  Robert De Niro            Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1    Jodie Foster            Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2   Albert Brooks                      Tom  ACTOR  Taxi Driver  MOVIE   
3   Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0

name      Character   r0le        TITLE   Type  release Year  \
0  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver  MOVIE          1976   
1    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver  MOVIE          1976   
2   Albert Brooks            Tom  ACTOR  Taxi Driver  MOVIE          1976   

               genres  imdb sc0re  imdb v0tes  
0  ['drama', 'crime']         8.2    808582.0  
1  ['drama', 'crime']         8.2    808582.0  
2  ['drama', 'crime']         8.2    808582.0

name      Character   r0le        TITLE   Type  release Year  \
0  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver  MOVIE          1976   
1    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver  MOVIE          1976   

               genres  imdb sc0re  imdb v0tes  
0  ['drama', 'crime']         8.2    808582.0  
1  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [8]:
# obtaining general information about the data in df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [9]:
# the list of column names in the df table
# Assuming df is your DataFrame
df.columns = [col.lower().replace(' ', '_').replace('0', 'o').strip() for col in df.columns]



Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [10]:
# renaming columns
df.columns = ['_'.join(col.split('_')).strip('_') for col in df.columns]

Check the result. Print the names of the columns once more:

In [11]:
# checking result: the list of column names
print(df.columns)

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    
Nice code! 

</div>

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [12]:
# calculating missing values
print(df.isnull().sum())

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64


Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [13]:
# dropping rows where columns with title, scores and votes have missing values
df['imdb_score'] = df['imdb_score'].dropna(0)  # or any other value you deem appropriate
df = df.dropna(subset=['imdb_score', 'imdb_votes'])
#changed to dropna correction

<div class="alert alert-info">
<h2> Student's comment</h2>

Correction made
</div>



<font color='red'> This code is used to change text color. </font>  

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
 
    
`fillna` does not drop anything. We are asked to remove these rows.

</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 </h2>
    

Good.     
    
</div>

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [14]:
# counting missing values
print(df.isnull().sum())

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64


### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [15]:
# counting duplicate rows
print(df.duplicated().sum())

6994


Review the duplicate rows to determine if removing them would distort our dataset.

In [16]:
# Produce table with duplicates (with original rows included) and review last 5 rows
df[df.duplicated(keep=False)].tail()
#no printing correction

name                 character   role    title   type  \
85569       Jessica Cediel           Liliana Navarro  ACTOR  Lokillo  MOVIE   
85570  Javier Gardeaz?­bal  Agust??n "Peluca" Ort??z  ACTOR  Lokillo  MOVIE   
85571        Carla Giraldo            Valery Reinoso  ACTOR  Lokillo  MOVIE   
85572  Ana Mar??a S?­nchez                   Lourdes  ACTOR  Lokillo  MOVIE   
85577         Isabel Gaona                    Cacica  ACTOR  Lokillo  MOVIE   

       release_year      genres  imdb_score  imdb_votes  
85569          2021  ['comedy']         3.8        68.0  
85570          2021  ['comedy']         3.8        68.0  
85571          2021  ['comedy']         3.8        68.0  
85572          2021  ['comedy']         3.8        68.0  
85577          2021  ['comedy']         3.8        68.0

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    

You don't need to call `print` here. 
</div>

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [17]:
# removing duplicate rows
df = df.drop_duplicates()

Check for duplicate rows once more to make sure you have removed all of them:

In [18]:
# checking for duplicates
print(df.duplicated().sum())

0


Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [19]:
# viewing unique type names
print(sorted(df['type'].unique()))


['MOVIE', 'SHOW', 'movies', 'shows', 'the movie', 'tv', 'tv series', 'tv show', 'tv shows']


Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [20]:
# function for replacing implicit duplicates
def replace_wrong_show(wrong_shows_list, correct_show):
    for show in wrong_shows_list:
        df['type'].replace(show, correct_show, inplace=True)
    # Iterate over the DataFrame and update 'type' column values directly
   
#correction to code

<div class="alert alert-info">
<h2> Student's comment</h2>

Correction made
</div>



<font color='red'> This code is used to change text color. </font>  

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
 
Try to not use a function inside another function. They are quite small, so there's no need in that. Would you try to update the code? 
    
</div>
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment 2 </h2>
    

Keep it simple:
</div>

```python

def replace_wrong_show(wrong_shows_list, correct_show):
    for show in wrong_shows_list:
        df['type'].replace(show, correct_show, inplace=True)
```

<div class="alert alert-info">
<h2> Student's comment 2</h2>

Replaced code with the one provided thank you for the correction
</div>



<font color='red'> This code is used to change text color. </font> 

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 3</h2>
    
Great!     
    
</div>

Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [21]:
# removing implicit duplicates
wrong_shows_list = ['shows', 'SHOW', 'tv show', 'tv shows', 'tv series', 'tv']
replace_wrong_show(wrong_shows_list, 'SHOW')
#correction to SHOW

<div class="alert alert-info">
<h2> Student's comment</h2>

Correction made
</div>



<font color='red'> This code is used to change text color. </font>  

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
We have to replace with `SHOW`. Next time, try to follow the recommendations, since it may be important. 
</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Great. You wrote and  called a function that worked properly.

In general, it can be performed without a function. To do this, we need a list and the same **replace** method. It will look like this:
    
    
    wrong_names = ['shows', 'tv show', 'tv shows', 'tv series', 'tv']
    correct_name = 'SHOW'  

    df['type'] = df['type'].replace(wrong_genres, correct_genre)
    
    
Or, which is the same:
    
    
    df['type'] = df['type'].replace(['shows', 'tv show', 'tv shows', 'tv series', 'tv'], 'SHOW')
</div>

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [22]:
# viewing unique genre names
print(sorted(df['type'].unique()))

['MOVIE', 'SHOW', 'movies', 'the movie']


### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [23]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name
df_filtered = df[df['release_year'] >= 1999]


In [24]:
# repeat conditional indexing so df has only shows (movies are removed as result)
df_filtered = df_filtered[df_filtered['type'] == 'SHOW']

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [25]:
# rounding column with scores

#checking the outcome with tail()
df_filtered['rounded_score'] = df_filtered['imdb_score'].round()
print(df_filtered.tail())

                      name character      role      title  type  release_year  \
85433     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  SHOW          2021   
85434    Rudklao Amratisha   unknown     ACTOR  Let's Eat  SHOW          2021   
85435      Jaturong Mokjok   unknown     ACTOR  Let's Eat  SHOW          2021   
85436     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  SHOW          2021   
85437  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  SHOW          2021   

                    genres  imdb_score  imdb_votes  rounded_score  
85433  ['drama', 'comedy']         8.2         5.0            8.0  
85434  ['drama', 'comedy']         8.2         5.0            8.0  
85435  ['drama', 'comedy']         8.2         5.0            8.0  
85436  ['drama', 'comedy']         8.2         5.0            8.0  
85437  ['drama', 'comedy']         8.2         5.0            8.0  


<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    
 
    
Always pay attention to the results. We have an empty dataframe because `show` was replaced with `SHOW`.
</div>

<div class="alert alert-info">
<h2> Student's comment 2</h2>

Found error corrected show with SHOW
</div>



<font color='red'> This code is used to change text color. </font>     

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 3</h2>
    
Correct :)     
</div>

It is now time to identify outliers based on the number of votes.

In [26]:
# Use groupby() for scores and count all unique values in each group, print the result
print(df_filtered.groupby('rounded_score')['imdb_votes'].nunique())
#correction with nunique function

rounded_score
2.0       5
3.0       4
4.0      30
5.0      91
6.0     329
7.0     568
8.0     527
9.0      65
10.0      1
Name: imdb_votes, dtype: int64


<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
We are asked about the unique values, not the number of rows. 
    

</div>

<div class="alert alert-info">
<h2> Student's comment</h2>

cannot find issue to show count. Switched to nunique
</div>



<font color='red'> This code is used to change text color. </font>      


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
`count` gives us the number of rows, not the number of unique values.

</div>

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [27]:
# filter dataframe using two conditions (scores to be in the range 4-9)

# group scores and corresponding average number of votes, reset index and print the result
df_filtered = df_filtered[df_filtered['rounded_score'].between(4, 9)]
average_votes = df_filtered.groupby('rounded_score')['imdb_votes'].mean().reset_index()
print(average_votes)
#result printed correction

   rounded_score     imdb_votes
0            4.0    5277.583333
1            5.0    3143.942568
2            6.0    3481.717322
3            7.0    8727.068211
4            8.0   30299.460967
5            9.0  126904.109515


<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    

It would be great if you printed the result :) 
</div>

<div class="alert alert-info">
<h2> Student's comment 2</h2>

Code is now working found error
</div>.



<font color='red'> This code is used to change text color. </font>  

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 3</h2>
    
Nice!     
</div>

Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [28]:
# round column with averages
average_votes['average_votes'] = average_votes['imdb_votes'].round()
# rename columns
average_votes = average_votes.rename(columns={'rounded_score': 'Score', 'imdb_votes': 'Average_Votes'})
# print dataframe in descending order
print(average_votes.sort_values(by='Average_Votes', ascending=False))

   Score  Average_Votes  average_votes
5    9.0  126904.109515       126904.0
4    8.0   30299.460967        30299.0
3    7.0    8727.068211         8727.0
0    4.0    5277.583333         5278.0
2    6.0    3481.717322         3482.0
1    5.0    3143.942568         3144.0


<div class="alert alert-info">
<h2> Student's comment 2</h2>

Made correction
</div>



<font color='red'> This code is used to change text color. </font>  

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    
The general conclusion is very important. Often, our tasks end with the  presentation or report for the client. Therefore, all results should be described in a general conclusion. Ideally, one should reinforce it with the previously obtained values.
</div>